[BEAM-456] Add MongoDbIO

jbonofre opened the pull request:
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [X] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [X] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [X] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [X] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---
This is the first "skeleton" for the MongoDbIO.

TODO:
- find a way to implement estimated size and split in the source
- check if it makes sense to manipulate `PCollection` of `Document` or `DBObject` instead of JSON `String`.

jkff commented:
Drive-by observations:
1) In case it turns out impractical to implement estimated size, progress computation, or dynamic splitting for Mongo, then you don't need the BoundedSource API and can implement the source as a simple ParDo. This still holds even if you support splitIntoBundles: it can be a separate ParDo with a fusion break in between. BoundedSource is only useful if you actually use its magical operations that ParDo's can't do, otherwise it's more difficult to develop against.
2) It may or may not be a good idea to add another fusion break after the read (e.g. a reshuffle), so that even though the reading happens in sequence, the processing would be parallel.

jbonofre commented:
Thanks for the feedback @jkff. Good point for BoundedSource vs ParDo. My purpose is to implement the BoundedSource feature. I'm experimenting different solutions. If it's not really possible, than I will refactore the MongDbIO Read to use ParDo internally instead of BoundedSource.

jbonofre commented:
By the way, the build was successful on Jenkins (regarding the console output).

jbonofre commented:
R: @dhalperi 
R: @jkff

jbonofre commented:
Rebased and first implementation of the {{getEstimatedSizeBytes()}} and {{splitIntoBundles()}} methods.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+ * <p>MongoDbIO source returns a bounded collection of String as {@code PCollection<String>}.
+ * The String is the JSON form of the MongoDB Document.</p>
+ *
+ * <p>To configure the MongoDB source, you have to provide the connection URI, the database name
```
Please wrap the long lines to 80 characters.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      return new Read(uri, database, collection, filter, numberSplit);
+    }
+
+    protected String uri;
```
Seems like all of these can be final.

jbonofre replied:
Agreed for private and final. For the source, I just mimic what we do in other IOs like Kafka, or JMS.

jbonofre replied:
Agreed for private. However, final is not possible there or it means I have to remove `Read()` and so users can't use `with*` setters or define null by default.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+    protected String collection;
+    @Nullable
+    protected String filter;
+    protected int numberSplit;
```
Is this specifying how many bundles to split into? Perhaps a better name, more consistent with the rest of Beam code style, would be "numSplits".

jbonofre replied:
Agreed. I'm changing that.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      org.apache.beam.sdk.io.Read.Bounded bounded = org.apache.beam.sdk.io.
+          Read.from(createSource());
+      PTransform<PInput, PCollection<String>> transform = bounded;
+      return input.getPipeline().apply(transform);
```
Can be a one-liner: return input.apply(org.apache.beam.sdk.io.Read.from(createSource());

jbonofre replied:
I think I need the cast to `PTransform` else I have a `Bounded`.

jkff replied:
Read.Bounded is a subclass of PTransform, otherwise line 157 wouldn't compile. Does the code not compile if you write it as I suggested?

jkff replied:
Read.Bounded is a subclass of PTransform, otherwise line 157 wouldn't compile. Does the code not compile if you write it as I suggested?

jbonofre replied:
Yes, it works, but I had to specify the type of the `BoundedSource`:

```
private static class BoundedMongoDbSource extends BoundedSource<String>
```

instead of

```
private static class BoundedMongoDbSource extends BoundedSource
```

jbonofre replied:
No:

```
incompatible types: org.apache.beam.sdk.values.POutput cannot be converted to org.apache.beam.sdk.values.PCollection<java.lang.String>
```

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      stat.append("collStats", collection);
+      Document stats = mongoDatabase.runCommand(stat);
+      totalCfByteSize = Long.valueOf(stats.get("size").toString());
+      avgSize = Long.valueOf(stats.get("avgObjSize").toString());
```
BoundedSource objects are not allowed to have mutable state; please make this a local variable.

jbonofre replied:
Good catch ! Thanks.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions pipelineOptions) {
+      Long estimatedByteSize = 0L;
```
Would be better to use a regular "long" - Long hints that these might be null, but that's not the case. Applies to all other usages of Long in this PR.

jbonofre replied:
Thanks, it makes sense.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+        try {
+          estimatedByteSize = getFilterResultByteSize(mongoCollection, avgSize);
+        } catch (Exception e) {
+          e.printStackTrace();
```
* Please use a logger rather than stderr.
* If we hit this branch, this function will implicitly return 0. Would be good to make this explicit (via "return 0;") if this is your intention.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+        throws Exception {
+      Long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
```
I'm not familiar enough with Mongo API, but is this a blocking call that executes the query and counts results? Seems like if the result dataset is large, this could be a significant sequential bottleneck. Please add some comments about what you're trying to accomplish here, either way.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      ArrayList splitIDFromMongo;
+      double initialRatio = 1.6;
+      if (numberSplit > 0) {
+        numberSplitsCalculated = new Long(numberSplit);
```
(same comments about usage of Long vs. long - prefer "long" everywhere, unless the value really can be null)

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      } else if (desiredBundleSizeBytes > 0) {
+        numberSplitsCalculated = collectionEstimatedSize / desiredBundleSizeBytes;
+        if (numberSplitsCalculated <= 0) {
+          numberSplitsCalculated = 1L;
```
Explain why in this case 1 is the proper result? E.g. you can get collectionEstimatedSize=0 if an error occurs while estimating the size - then is 1 still the proper split granularity to choose?

Either way, please add some comments explaining the reasoning, and ideally some logging too. As a user of this class, I would appreciate knowing how much parallelism the connector chose or was able to achieve.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+        listEntireCollection.add(this);
+        return listEntireCollection;
+      }
+      int maxChunkSize = 64000;
```
* What are the units? (ideally put in the variable name - seems like it's bytes, so maxChunkSizeBytes is a good name)
* Explain why is 64000 a reasonable value?

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      Document data = mongoDatabase.runCommand(split);
+      splitIDFromMongo = (ArrayList) data.get("splitKeys");
+      // ratio between number of splits from mongo and splits defined by runner or user
+      double calculatedRatio = (double) splitIDFromMongo.size() / numberSplitsCalculated;
```
It would be good to extract the code of this loop into a utility function with its own unit tests, since it seems to not require MongoDB interaction, but can have tricky floating-point edge cases.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+          && (splitIDFromMongo.size() > numberSplitsCalculated))) {
+        ArrayList splitIDCalculated = new ArrayList();
+        double mod = splitIDFromMongo.size() % numberSplitsCalculated;
+        HashMap<Integer, Integer> idRanges = new HashMap<Integer, Integer>();
```
Seems like this can be an array rather than a hashmap: the code isn't doing any lookups, or even accessing the keys, and the keys are a dense range [0, numberSplitsCalculated).

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+        // compute the ranges from the split Vector command
+        Object splitId = 0L;
+        int indexPosition = 0;
+        // isPreviousTwo is used to managed the change of value for the range (from 2 to 1)
```
I'm lost here and don't quite understand what the code is doing. Could you add a comment giving some sort of example?

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      String lastID = null; // Lower boundary of the first min split
+      int listIndex = 0;
+      for (final Object splitIdValue : splitIDList) {
+//          final Document  currentDoc = (Document) aSplitData;
```
Remove commented-out code.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      return createSourceList(splitIDFromMongo);
+    }
+
+    private List<BoundedSource> createSourceList(ArrayList splitIDList) {
```
Seems like it should be possible to make this function static and have its own unit tests.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+  private static class BoundedMongoDbReader extends BoundedSource.BoundedReader {
+
+    private String uri;
```
Rather than listing the configuration variables, can you store a single "private final BoundedMongoDbSource source"? That will be fewer code and easier to maintain.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+    @Override
+    public BoundedSource getCurrentSource() {
+      return null;
```
This is not correct: getCurrentSource() must return the source from which this reader was created (modulo splitAtFraction() effects, but your reader doesn't use it). This should be trivial to do if you store the source in the constructor.

jbonofre replied:
Good point, thanks. I'm fixing that.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+    @Override
+    public void close() {
+      cursor.close();
```
Seems like if cursor.close() throws an exception, then you won't call client.close(). Is this safe?

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+      return new Write(uri, database, collection);
+    }
+
+    protected String uri;
```
Same suggestion about avoiding duplication of configuration variables.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+        MongoDatabase mongoDatabase = client.getDatabase(database);
+        MongoCollection mongoCollection = mongoDatabase.getCollection(collection);
+
+        mongoCollection.insertOne(Document.parse(value));
```
Seems like this could be dramatically more efficient with batching (accumulate a batch in processElement(), and flush it by a multi-insert - I suppose Mongo supports multi-inserts - either when the batch grows too big or in finishBundle). Is this something you plan to do?

jbonofre replied:
Correct, MongoDB supports insertAny(), and it's on my TODO ;)

jkff replied:
I think it'd make a lot of sense to do this in the current PR, because I expect the performance difference to be dramatic (i.e. performance of the current code might be surprisingly low, unexpectedly for someone who tries to use this connector with a large amount of data - e.g. here http://info.meteor.com/blog/inserting-50000-documents-into-a-collection-slow-fast-and-fastest the difference appears to be ~10x). Batching doesn't seem hard to implement (though there's a long-standing feature request somewhere to make it easier to implement batching DoFn's...).

jbonofre replied:
All right, I'm implementing sink batching then ;)

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+      @StartBundle
+      public void startBundle(Context c) throws Exception {
+        if (client == null) {
```
Can client be non-null at this point?

jbonofre replied:
I think it could be null when starting the bundle for the first time.

jkff replied:
Please use the @Setup and @Teardown methods to create and destroy the client. That's more efficient, because they're invoked just once per DoFn instance, and then you don't need null checks (it will *definitely* be null in @Setup, and *definitely* non-null in @Teardown).

jkff commented:
Thanks JB! Sorry for the delay in review.

jbonofre commented:
Thanks to you !! I will implement your comments. Thanks again.

jbonofre commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,535 @@
+
+  private static class BoundedMongoDbSource extends BoundedSource {
+
+    private String uri;
```
Agreed.

jbonofre commented:
Rebased and implementing Eugene's comments.

amarouni commented:
@jbonofre 
Does it make any sense to add DDL (drop existing collection, ...) options to the IO ?
(These would be executed once the IO is instantiated on the driver side.)
Or it makes more sense to keep these outside of the IO ?

jbonofre commented:
@amarouni it's a good idea. My plan is first to complete this PR (the split and estimated size should be improved, I'm working on it). The DDL option could be part of a new PR. WDYT ?

amarouni commented:
@jbonofre Great

jbonofre commented:
Rebased and updated with:
- refactoring of the splitIntoBundle() using the desiredBundleSize provided by the runner or calculated using the numSplit provided by user
- reuse attributes from Source or Write in the Reader and Writer

@jkff can we discuss about it together on Hangout when you have a chance ? Thanks !

jbonofre commented:
Weird, Jenkins build looks good:

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 22:53 min
[INFO] Finished at: 2016-08-29T10:09:29+00:00
[INFO] Final Memory: 189M/1028M
[INFO] ------------------------------------------------------------------------
```

but it seems that the tests failed:

```
com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=UNKNOWN, servers=[{address=0.0.0.0:27017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketOpenException: Exception opening socket}, caused by {java.net.SocketTimeoutException: connect timed out}}]
	at com.mongodb.connection.BaseCluster.createTimeoutException(BaseCluster.java:369)
```

I'm checking why.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+ * pipeline.apply(MongoDbIO.read()
+ *   .withUri("mongodb://localhost:27017")
+ *   .withDatabase("my-database")
+ *   .withCollection("my-collection")
```
Close the brace opened at line 67: );

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+ *     .withUri("mongodb://localhost:27017")
+ *     .withDatabase("my-database")
+ *     .withCollection("my-collection")
+ *     .withNumSplits(30)
```
Here too.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    private final String uri;
```
I'd like to again suggest making the Read class *not* store the variables individually. Instead do like this:
class Read {
  private final BoundedMongoDbSource source;
  public Read() { this.source = new BoundedMongoDbSource(null, null, null, null, 0); }
  public Read withUri(String uri) { return new Read(source.withUri(uri)); }
  public Read withDatabase(String database) { return new Read(source.withDatabase(database)); }
  ...
}

And move your current withBlah methods into BoundedMongoDbSource.

jbonofre replied:
Yes good idea, largely better indeed. I will do this way. Thanks !

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+    }
+
+    @Override
+    public void validate(PBegin input) {
```
This should simply validate to source.validate(), given the suggestion above.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+    public void populateDisplayData(DisplayData.Builder builder) {
+      super.populateDisplayData(builder);
+
+      builder.addIfNotNull(DisplayData.item("uri", uri));
```
This should also delegate to the source.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbIO.class);
+
+  public static Read read() {
```
Javadoc all public classes and all their public members. For brevity, ok to use /** blah */ rather than multi-line style.
Make sure to document carefully the things that have subtle semantics, e.g. numSplits.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+      // filter using the average object size * the number of documents
+      long totalFilterByteSize = 0L;
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
```
Same question as last time: does this actually execute the query against the whole database? It sounds like a sequential bottleneck that could defeat the parallelism you're achieving in splitIntoBundles.

jbonofre replied:
It does query the database, but it doesn't retrieve the documents (results). This method is used to estimate the size of the filter (provided by user): number of documents * average size of document. An alternative would be to not consider the filter at all in the estimated size and just consider the collection.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+      Document bson = Document.parse(filter);
+      long countFilter = mongoCollection.count(bson);
+      totalFilterByteSize = countFilter * avgSize;
+      return totalFilterByteSize;
```
This whole method can be abbreviated to

return mongoCollection.count(Document.parse(filter)) * avgSize;

and at this point you don't need the method at all - getting rid of 10 lines of code.

jbonofre replied:
+1 Thanks

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+      splitVectorCommand.append("keyPattern", new BasicDBObject().append("_id", 1));
+      splitVectorCommand.append("force", false);
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
```
* What will this do if desiredBundleSizeBytes is zero or small?
* Please add logging about how many parts you actually split into. It will be valuable for debugging when a pipeline is slow and it is because the source split into very few parts.

jbonofre replied:
Good point: if desiredBundleSizeBytes is 0, than I will create an unique source. Good idea for the logging.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+      // maxChunkSize is the Mongo partition size in MB
+      splitVectorCommand.append("maxChunkSize", desiredBundleSizeBytes / 1024 / 1024);
+      Document splitVectorCommandResult = mongoDatabase.runCommand(splitVectorCommand);
+      splitKeys = (ArrayList) splitVectorCommandResult.get("splitKeys");
```
* Can you use a generic ArrayList rather than the raw type? Or is the result actually a list of heterogeneously typed Object's so you can't cast it to, say, ArrayList<String>?
* Is the result *guaranteed* to be an ArrayList rather than, say, some other subtype of List or Collection - i.e. is this a guarantee explicitly advertised by the MongoDb client library?

jbonofre replied:
I gonna check but AFAIR we can use a ArrayList: splitKeys is always a list of keys.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+     */
+    private List<BoundedSource> createSourceList(ArrayList splitKeys) {
+      List<BoundedSource> splitSourceList = new ArrayList<>();
+      if (splitKeys == null) {
```
Better make this method take a guaranteed non-null argument, and do this null checking at the call site. This will make the invariants of the current method stricter and hence easier to understand, and will also reduce nesting in code.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+        splitSourceList.add(new BoundedMongoDbSource(uri, database, collection, filter, numSplits));
+      } else {
+        String lastID = null; // lower boundary of the first min split
+        int listIndex = 0;
```
Seems like the loop below should be an indexed loop:
for (int i = 0; i < splitKeys.size(); ++i) {
  String splitKey = splitKeys.get(i).toString();
  ...
}

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+          String filterWithBoundaries;
+          if (filter != null && !filter.isEmpty()) {
+            if (listIndex == 0) {
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$lte:ObjectId(\""
```
I think all these string operations would read better (and easier to check for correctness) using String.format() rather than explicit concatenation.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+                  + "\")," + "$lte:ObjectId(\"" + currentID + "\")}}, "
+                  + filter + " ]}";
+            }
+          } else {
```
Seems like you can build a string for the ObjectId filter, and then optionally also wrap it into an $and with the specified filter (if any). That way you'll have fewer code duplication between lines 317-332 and 334-345.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+              filterWithBoundaries = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                  + "\")," + "$lt:ObjectId(\"" + currentID
+                  + "\")}}, " + filter + " ]}";
+              splitSourceList.add(new BoundedMongoDbSource
```
I'm confused, why does this branch add something to the list but others don't? Is there a way to structure the code in a way that none of the branches added something, but simply constructed a filter string?

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+                  + "$lte:ObjectId(\"" + currentID + "\")}}";
+            }
+          }
+          splitSourceList.add(new BoundedMongoDbSource
```
I'd suggest to encapsulate the logic of this loop (converting a list of split keys into a list of filters) into a utility function, and separately invoke it to produce the actual BoundedMongoDbSource objects:

private static List<String> splitKeysToFilters(List<String> splitKeys, String additionalFilter) {
  ...
}

splitIntoBundles() {
  for (String shardFilter : splitKeysToFilters(splitKeys, filter)) {
    res.add(this.withFilter(shardFilter));
  }
}

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+    @Override
+    public void close() {
+      try {
+        cursor.close();
```
if (cursor != null)

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,500 @@
+        client.close();
+        client = null;
+      }
+
```
These extra blank lines are unnecessary.

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -0,0 +1,167 @@
+    MongoDatabase database = client.getDatabase(DATABASE);
+
+    MongoCollection collection = database.getCollection(COLLECTION);
+    collection.drop();
```
Why is it necessary to drop the collection: doesn't the database start off empty?

jbonofre replied:
Agree, just habit ;) Removing this.

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -0,0 +1,167 @@
+          .withCollection(COLLECTION));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally()))
+        .isEqualTo(new Long(1000));
```
Add verification that the scientists are mentioned the expected number of times? (e.g. if you use 8 scientists, each will be mentioned exactly 125 times, right?)

jbonofre replied:
Good idea, I will add the corresponding `PAssert`.

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -0,0 +1,167 @@
+        .withFilter("{\"scientist\":\"Einstein\"}"));
+
+    PAssert.thatSingleton(output.apply("Count", Count.<String>globally()))
+        .notEqualTo(new Long(1000));
```
This assertion is too weak and it could pass due to a bug in the code rather than due to correct behavior - it'd be better to use an exact value.

jbonofre replied:
Good point, I'm using exact value.

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -0,0 +1,167 @@
+
+    pipeline.run();
+
+    MongoClient client = new MongoClient("localhost", PORT);
```
You could read data using MongoDbIo.read() and verify using PAssert :)

jbonofre replied:
As I would like to test `write` here, I prefer to use core MongoDB mechanism to validate the test. I think it's more secure and reliable (for instance if there's an issue in `MongoDbIO.read()`. WDYT ?

jkff replied:
Seems reasonable.

jbonofre commented:
Rebased and updated.

jkff commented on sdks/java/io/mongodb/pom.xml:
```diff
@@ -0,0 +1,129 @@
+    </dependency>
+    <dependency>
+      <groupId>org.apache.beam</groupId>
+      <artifactId>beam-runners-direct-java</artifactId>
```
Just wondering: shouldn't these be @RunnableOnService tests, subject to running on all runners? (I'm not familiar with how that part of our testing works yet)

jbonofre replied:
It is, but by default, I have to define at least the direct pipeline runner for the test (else there's no runner by default in the classpath).

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,539 @@
+     * Creates a {@link BoundedSource} with the configuration in {@link Read}.
+     */
+    @VisibleForTesting
+    BoundedSource<String> createSource() {
```
Should probably be called getSource(), or just make the variable itself @VisibleForTesting.

jbonofre replied:
Right, let me do the change.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,539 @@
+      builder.addIfNotNull(DisplayData.item("uri", uri));
+      builder.addIfNotNull(DisplayData.item("database", database));
+      builder.addIfNotNull(DisplayData.item("collection", collection));
+      builder.addIfNotNull(DisplayData.item("filter", filter));
```
Of these, only filter can be null, right?

jbonofre replied:
Correct (according to `validate()`).

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,539 @@
+        // by counting number of document returned by the filter
+        // and using the average object size from collStats
+        try {
+          estimatedSizeBytes = mongoCollection.count(Document.parse(filter)) * avgSize;
```
Once again: mongoCollection.count() does a full scan of the table, which will be extremely slow for large tables (MongoDB website boasts deployments that store 100B+ documents), which defeats the whole purpose of BoundedSource (reading independent parts of the table in parallel). Please don't do it. The user's pipeline will be bottlenecked for a long time doing getEstimatedSize(), followed of course by a slightly better-tuned parallel scan of the actual data, but it'll be too late. Just ignore the filter for purposes of estimating size; and when we see behavior that says ignoring the filter caused serious issues, let's deal with it on a case-by-case basis.

By the way, is the collStats() command O(1)?

jbonofre replied:
`count()` use the filter, so it means it's will count the number of document returned by the filter.

Yes, `collStats` doesn't iterate on the collection, it directly gets the statistics on the collection (size and average document size in the collection). It's pretty straight forward (no computation).

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,539 @@
+      }
+
+      // the desired bundle size is small, using default chunk size of 64MB
+      if (desiredBundleSizeBytes == 0 || ((desiredBundleSizeBytes / 1024 / 1024) < 1)) {
```
This can be simplified to "if (desiredBundleSizeBytes < 1024 * 1024)". And then the condition looks weird: why do we adjust desiredBundleSizeBytes under 1MB to 64MB, but, say, 3MB we'll keep as is? Shouldn't this be just desiredBundleSizeBytes = max(desiredBundleSizeBytes, 64MB) ? (I don't know whether or not this is the right thing to do)

jbonofre replied:
My point was more to allow user to define a number of splits. If we do `max(desiredBundleSizeBytes, 64MB)`, it won't be the number of splits defined by the user.
I will improve by:
- if the `desiredBundleSizeBytes < 1024 * 1024`, then use 1MB as split
- else use the calculated `desiredBundleSizeBytes`

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -0,0 +1,539 @@
+          rangeFilter = String.format("{ $and: [ {\"_id\":{$lte:Objectd(\"%s\")}}",
+              splitKey);
+        }
+        if (i == splitKeys.size() - 1) {
```
You can rewrite these 3 mutually exclusive conditions to a slightly simpler form:
if (i == 0) {
...
} else if (i == splitKeys.size() - 1) {
...
} else {
...
}

jbonofre replied:
I did this way because I thought it was easier to read. Let me do as I did initially.

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -0,0 +1,217 @@
+
+    PAssert.thatSingleton(
+        output.apply("Filter Einstein", Filter.by(new ScientistFilter("Einstein")))
+        .apply("Count Einstein", Count.<String>globally())).isEqualTo(100L);
```
I'm not sure why this code duplication is necessary. Can you convert this to a loop?

jbonofre replied:
Good idea ;) Let me do that.

jkff commented:
Thanks, this is starting to look pretty awesome! The only substantial comment is about performance of count() - other than that, this is likely the last round of cosmetic comments.

jkff commented:
Also, I'm not sure why but the Jenkins thing is failing...

jbonofre commented:
Updated.

jbonofre commented:
I'm checking why Jenkins is failing.

jbonofre commented:
Oh, I think I found the cause of the issue: embedded MongoDB uses `/tmp` to store the artifacts. I think Jenkins don't allow this. I'm changing the location of the MongoDB storage to use the `target` folder and give a try.

jbonofre commented:
Testing if it works on Jenkins with the MongoDB instance located in the target folder.

jkff commented on sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java:
```diff
@@ -519,14 +507,24 @@ public void createMongoClient() throws Exception {
       public void processElement(ProcessContext ctx) throws Exception {
         String value = ctx.element();
 
+        batch.add(Document.parse(ctx.element()));
```
This will OOM if you have a bundle with a lot of elements (more likely in a batch pipeline, but can happen in a streaming pipeline too). You need to limit the size of the batch and flush it both when it reaches the limit and in finishBundle().

jkff commented on sdks/java/io/mongodb/src/test/java/org/apache/beam/sdk/io/mongodb/MongoDbIOTest.java:
```diff
@@ -113,56 +118,27 @@ public void testFullRead() throws Exception {
+    ).satisfies(new SerializableFunction<Iterable<KV<String, Long>>, Void>() {
+      @Override
+      public Void apply(Iterable<KV<String, Long>> input) {
+        Iterator<KV<String, Long>> iterator = input.iterator();
```
Use a regular Java foreach loop: for (KV<String, Long> element : input) { assertEquals(100L, element.getValue()); }

jbonofre replied:
+1

jbonofre replied:
+1

jbonofre commented:
Tests are still failed on Jenkins (whereas they work fine on my machine):

```
com.mongodb.MongoTimeoutException: Timed out after 30000 ms while waiting for a server that matches WritableServerSelector. Client view of cluster state is {type=UNKNOWN, servers=[{address=0.0.0.0:27017, type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSocketOpenException: Exception opening socket}, caused by {java.net.SocketTimeoutException: connect timed out}}]
```

I saw in the log that a socket file is open in `/tmp` folder, it could be related. I'm checking.

jkff commented:
I think you need to use a @TemporaryFolder rule to get the tmp location http://stackoverflow.com/questions/12400243/what-is-the-correct-way-to-write-to-temp-file-during-unit-tests-with-maven

Other than that - please fix the batching in the sink, and I think it'll be good to go! Thanks for bearing with me!

jbonofre commented:
Thanks, I'm doing the changes ! Much appreciated !

jbonofre commented:
I'm not sure for @TemporaryFolder: the problem is that the EmbeddedMongoDB uses java.io.tmpdir by default. So, I changed that to the storage location (and use target). Let me check for the socket file.

jbonofre commented:
Updated, and I fixed the tests for Jenkins (forcing the binding to `localhost` instead of `0.0.0.0`).

jkff commented:
LGTM, thank you! This can pretty much be used as an exemplar connector implementation for future authors now!

jbonofre commented:
Thanks buddy, I gonna merge on master ! :+1:
